### ISSUE TITLE
It is no longer working.

### ISSUE THREAD

**Issue by Mikasa-san** [avi12/youtube-downloader#2]
Please update it. It was working yesterday.

**Comment by avi12** [comment]
It seems that the playlist downloads got broken

I also noticed that no matter the browser I'm running it on (Chrome, Edge, Opera), they all mention the "Manifest version 2 will not be supported starting from January 2023"
This sadly cannot be fixed. Instead, someone needs to rethink the extension's architecture to port it to [Manifest V3](https://developer.chrome.com/docs/extensions/mv3/intro/mv3-overview), and I definitely can't do this alone, and weekends are definitely not enough (as I have a full-time job)

**Comment by guest271314** [comment]
> This sadly cannot be fixed. Instead, someone needs to rethink the extension's architecture to port it to [Manifest V3](https://developer.chrome.com/docs/extensions/mv3/intro/mv3-overview), and I definitely can't do this alone, and weekends are definitely not enough (as I have a full-time job)

What needs to be done?

**Comment by avi12** [comment]
As I said, a complete rethinking of the way the extension works
I'm not even sure if some of the things will work in the first place
For example, because a Service Worker is no longer a web page like the background page used to be, this line will fail:
https://github.com/avi12/youtube-downloader/blob/8a42311323d2557bd7a3666c28b438f802e1df22/src/scripts/background.ts#L318
Which is basically a backbone, as the [URL Web API](https://developer.mozilla.org/en-US/docs/Web/API/URL) is necessary to download the actual file
A possible workaround is to send that URL to the content script and then trigger a download via [`file-saver`](https://www.npmjs.com/package/file-saver)

Next up:
1. The extension currently relies on [FFmpeg](https://github.com/ffmpegwasm/ffmpeg.wasm) to combine the video & audio (as YouTube provides them separately)
   1. I don't know whether FFmpeg can work in a Service Worker
   2. If it does, you'll have to take into account the [5-minute time limit](https://chromium.googlesource.com/chromium/src/+/master/docs/security/service-worker-security-faq.md#do-service-workers-live-forever) of Service Workers
      This means: If the user wants to download a large video (lengthy and/or high quality), you need to make sure to first download the video & audio, probably wait until the Service Worker is killed (I don't know how to check such a thing), and then run an FFmpeg job to combine them, yet - depending on the user's hardware, it might take more than 5 minutes, in which case you'll need to show an error
      Of course, make sure to allow the user to cancel the download at any time
1. Regarding file downloads, you'll probably want to use [Background Fetch](https://developer.chrome.com/blog/background-fetch/)
   - Make sure that you can calculate a download's progress, as well as cancel it at any time
   - After the download is complete, you can potentially store it in the storage as base64 and then fetch later to use for the FFmpeg job

Bonus: Try to support Firefox, as it currently faces the `SharedArrayBuffer` [issue](https://github.com/ffmpegwasm/ffmpeg.wasm/issues/263)

Good luck!
Since I have a full-time job, I can't guarantee to be available, but I can try to help in some stuff

**Comment by guest271314** [comment]
> As I said, a complete rethinking of the way the extension works I'm not even sure if some of the things will work in the first place For example, because a Service Worker is no longer a web page like the background page used to be, this line will fail:
> 
> https://github.com/avi12/youtube-downloader/blob/8a42311323d2557bd7a3666c28b438f802e1df22/src/scripts/background.ts#L318
> 
> 
> Which is basically a backbone, as the [URL Web API](https://developer.mozilla.org/en-US/docs/Web/API/URL) is necessary to download the actual file

We can send `dataFile` to any other Web page using `postMessage()`.

> A possible workaround is to send that URL to the content script and then trigger a download via [`file-saver`](https://www.npmjs.com/package/file-saver)

File System Access API can be used on Chromium/Chrome.

> Next up:
> 
> 1. The extension currently relies on [FFmpeg](https://github.com/ffmpegwasm/ffmpeg.wasm) to combine the video & audio (as YouTube provides them separately)
>    
>    1. I don't know whether FFmpeg can work in a Service Worker

Yes.

Alternative approaches include using Native Messaging and FFmpeg installed on the users machine.

>    2. If it does, you'll have to take into account the [5-minute time limit](https://chromium.googlesource.com/chromium/src/+/master/docs/security/service-worker-security-faq.md#do-service-workers-live-forever) of Service Workers
>       This means: If the user wants to download a large video (lengthy and/or high quality), you need to make sure to first download the video & audio, probably wait until the Service Worker is killed (I don't know how to check such a thing), and then run an FFmpeg job to combine them, yet - depending on the user's hardware, it might take more than 5 minutes, in which case you'll need to show an error
>       Of course, make sure to allow the user to cancel the download at any time

There are several ways to keep a `ServiceWorker` active indefinitely, see https://github.com/guest271314/persistent-serviceworker.


> 2. Regarding file downloads, you'll probably want to use [Background Fetch](https://developer.chrome.com/blog/background-fetch/)
>    
>    * Make sure that you can calculate a download's progress, as well as cancel it at any time
>    * After the download is complete, you can potentially store it in the storage as base64 and then fetch later to use for the FFmpeg job
> 
> Bonus: Try to support Firefox, as it currently faces the `SharedArrayBuffer` [issue](https://github.com/ffmpegwasm/ffmpeg.wasm/issues/263)
> 

We can remove `Content-Security-Policy` headers altogether, see https://github.com/guest271314/remove-csp-header. Where once removed it is possible to fetch chrome-extension: URL's or localhost from any Web page.

> Good luck! Since I have a full-time job, I can't guarantee to be available, but I can try to help in some stuff

I do, too. I am trying to help where I am able. I have not used your specific code. I have used [youtube-dl](https://youtube-dl.org/). The deprecation of MV2 and transition to MV3 can be worked around.

**Comment by avi12** [comment]
@guest271314 I forgot to mention that this needs to be worked around as well:
https://github.com/avi12/youtube-downloader/blob/f6ff4f64f1f32b68c0d83d17f54f277d522a5642/src/scripts/yt-downloader-functions.ts#L41-L48
I went back to [ytdlr](https://github.com/bakapear/ytdlr/blob/master/ytdlr.js), and apparently, the maintainer has updated the code to use YouTube's API, rather than relying on `eval`, which is great news

> Alternative approaches include using Native Messaging and FFmpeg installed on the user's machine

Not good. I wish the extension to be standalone and not rely on the user installing anything

> We can remove `Content-Security-Policy` headers altogether, see https://github.com/guest271314/remove-csp-header. Where once removed it is possible to fetch chrome-extension: URL's or localhost from any Web page.

That sounds very interesting. I'm curious to see how it can be applied in practice

> I have used [youtube-dl](https://youtube-dl.org/).

Me too. By debugging the code, I realized very important aspects that resulted in me seeking the decipher function I pasted a snippet of at the beginning of this comment

**Comment by guest271314** [comment]
> @guest271314 I forgot to mention that this needs to be worked around as well:
> 
> https://github.com/avi12/youtube-downloader/blob/f6ff4f64f1f32b68c0d83d17f54f277d522a5642/src/scripts/yt-downloader-functions.ts#L41-L48

I don't know what that function does.

> I went back to [ytdlr](https://github.com/bakapear/ytdlr/blob/master/ytdlr.js), and apparently, the maintainer has updated the code to use YouTube's API, rather than relying on `eval`, which is great news
> > Alternative approaches include using Native Messaging and FFmpeg installed on the user's machine
> 
> Not good. I wish the extension to be standalone and not rely on the user installing anything

I'm not sure what you mean?


The user has to install FFmpeg.wasm. If FFmpeg is already installed on their machine no installation of anything else is necessary to utilize Native Messaging to communicate with the already installed application, e.g., I use already installed `mkvmerge` here https://github.com/guest271314/native-messaging-mkvmerge and already installed eSpeak NG here https://github.com/guest271314/native-messaging-espeak-ng.

> > We can remove `Content-Security-Policy` headers altogether, see https://github.com/guest271314/remove-csp-header. Where once removed it is possible to fetch chrome-extension: URL's or localhost from any Web page.
> 
> That sounds very interesting. I'm curious to see how it can be applied in practice

Simply put, you can `fetch()` resources from a local server or chrome-extension: protocol URL's from any Web page. If that functionality is necessary. For localhost just serve with the following headers:

`Access-Control-Allow-Headers: Access-Control-Request-Private-Network`
`Access-Control-Allow-Private-Network: true`
 
> > I have used [youtube-dl](https://youtube-dl.org/).
> 
> Me too. By debugging the code, I realized very important aspects that resulted in me seeking the decipher function I pasted a snippet of at the beginning of this comment

**Comment by avi12** [comment]
@guest271314 I'm starting to investigate what can be done for the download part and so far I'm getting negative results as regards [Background Fetch](https://developer.chrome.com/blog/background-fetch/), i.e. when I initialize a download for a video, it will keep failing

**Comment by guest271314** [comment]
The last time I checked Background Fetch is not implemented in MV3 `ServiceWorker` for extension https://github.com/mdn/content/issues/12370.

You can use a proxy to `fetch()` external resources from inside to outside of the `ServiceWorker` if the issue is CORS restrictions.

**Comment by guest271314** [comment]
What happens when you just use `fetch()`?

**Comment by avi12** [comment]
I believe [this](https://github.com/w3c/ServiceWorker/issues/1188) is the relevant issue

**Comment by avi12** [comment]
> What happens when you just use `fetch()`?

I can use `fetch` as usual, but the issue is that a Service Worker in the MV3 extension has a 5-minute lifespan limit
Imagine wanting to download a 1-hour podcast, only to be interrupted because Chrome killed the Service Worker
The Background Fetch API was supposed to be the solution, but so far I cannot manage to download external files, so it's irrelevant

**Comment by guest271314** [comment]
There are workarounds [for that ](https://github.com/guest271314/persistent-serviceworker).

**Comment by guest271314** [comment]
I have kept the same MV3 `ServiceWorker` active for well over 24 hours using the approaches described here https://github.com/guest271314/persistent-serviceworker.

**Comment by avi12** [comment]
Did you use this one? https://github.dev/guest271314/persistent-serviceworker/blob/c97f3092b099c33d43a86d1e2add176238e1cc3b/chromium_extension_web_accessible_resources_iframe_message_event

**Comment by guest271314** [comment]
All of them.

**Comment by guest271314** [comment]
If any of the approaches do not work as expected kindly file an issue.

**Comment by avi12** [comment]
Right, but what would be the best fit for this extension's use case?

**Comment by guest271314** [comment]
Any one. Message passing between an `<iframe>` and `ServiceWorker` works https://github.com/guest271314/persistent-serviceworker/tree/main/chromium_extension_web_accessible_resources_iframe_message_event. That is what I use to stream and record hours of live audio. Technically any never ending fetch request works.

**Comment by guest271314** [comment]
I suggest testing all of the approaches. Then make a decision based on evidence.

**Comment by guest271314** [comment]
> Imagine wanting to download a 1-hour podcast

It shouldn't take long to _download_ a 1 hour podcast. I fetch Node.js binary download that includes `node` executable, `npm`, etc. in a few seconds.

**Comment by avi12** [comment]
Basically, I need to _fetch_ the podcast, both the video and the audio, _send_ both over to FFmpeg to be merged into a single MP4 (or any other video format, for that matter), and then _initialize the download_ on the user's end so he can both of view the download progress and cancel it at any time

**Comment by guest271314** [comment]
All possible. The `ServiceWorker` becoming inactive can be worked around using any of the approached linked above. The download can use `fetch()` piped to a File System Access API `FileSystemWritableFileStream` with `AbortController` set.

**Comment by avi12** [comment]
Sounds great, I will investigate the approaches mentioned in your repo in a few hours

**Comment by avi12** [comment]
I tried testing `chromium_extension_web_accessible_resources_iframe_message_event`, but clicking on the extension icon threw an error:
```
Uncaught (in promise) Error: Cannot access a chrome:// URL
```
at [this line](https://github.com/guest271314/persistent-serviceworker/blob/main/chromium_extension_web_accessible_resources_iframe_message_event/background.js#L67)

**Comment by guest271314** [comment]
Does the manifest.json include the following?

```
"host_permissions": ["<all_urls>"],
```

**Comment by avi12** [comment]
Yes, I cloned the repo and tried sideloading the extension as-is

**Comment by guest271314** [comment]
Do you really need to run the code on a `chrome:` URL?

**Comment by avi12** [comment]
I just cloned and sideloaded the repo to see that it works and I don't get errors

**Comment by guest271314** [comment]
What happens when you run the code on a URL that is not `chrome:` protocol?

**Comment by guest271314** [comment]
It will require further code to access a `chrome:` URL or `chrome-extension:` from another extension; i.e., modify or remove CSP headers and explicitly set the chrome: URL you want to access. I think you are testing on a chrome: protocol URL, which is probably not what users will do. You can create an index.html page in the extension for users to use if that is an option you want to provide. By default chrome extensions cannot access chrome URL's or other extensions. When you navigate to chrome://flags you will note the flag 

> Extensions on chrome:// URLs
>
> Enables running extensions on chrome:// URLs, where extensions explicitly request this permission. – Mac, Windows, Linux, ChromeOS, Android, Fuchsia, Lacros
>
> [#extensions-on-chrome-urls](chrome://flags/#extensions-on-chrome-urls)

**Comment by guest271314** [comment]
See https://groups.google.com/a/chromium.org/g/chromium-extensions/c/9uC1EiIxMlQ/m/F-ElRDkRAgAJ

> 2) the focused tab was on a page that does not allow script injection. Notably these include Chrome pages (chrome://) and extension pages (chrome-extension://)

**Comment by avi12** [comment]
Ok, that's useful info, but I'm not sure what's my solution

**Comment by guest271314** [comment]
Solution for what? As long as you don't try to inject the script on pages such as `chrome-extensions` or `chrome://settings` the `ServiceWorker` establish communication with the Tab and remain actie indefinitely,

**Comment by avi12** [comment]
Can you make a quick repo example of having a constant fetch to a certain site?
I didn't fully understand how to approach it programmatically

**Comment by guest271314** [comment]
I don't think you need a constant fetch. However, that is what this does using `EventSource` https://guest271314.github.io/persistent-serviceworker/ and this does with `fetch()` and `respondWith()` https://github.com/guest271314/persistent-serviceworker/tree/main/readablestream-fetch-respondwith.

How do you fetch audio and video from YouTube?

**Comment by avi12** [comment]
> How do you fetch audio and video from YouTube?

```ts
const url = "https://www.youtube.com/watch?v=u4nkpIyPkSE";
const text = await (await fetch(url)).text();
const match = text.match(/ytInitialPlayerResponse\s*=\s*({.+?})\s*;\s*(?:var\s+meta|<\/script|\n)/)[1];
const json = JSON.parse(match);
const { adapticeFormats } = json.streamingData;
const videoObj = adaptiveFormats.find(({ mimeType }) => mimeType.startsWith("video"));
const audioObj = adaptiveFormats.find(({ mimeType }) => mimeType.startsWith("audio"));
const urlVideo = videoObj.url;
const urlAudio = audioObj.url;

const [dataVIdeo, dataAudio] = await Promise.all([fetch(urlVideo), fetch(urlAudio)]);
```
The last line has to be a little bit more complex, has to include features such as progress tracking and the ability to cancel the fetching, but that's the gist of it
To combine them, I need to use a program like FFmpeg, then start the download using [chrome.downloads](https://developer.chrome.com/docs/extensions/reference/downloads/#method-download)

**Comment by avi12** [comment]
> I don't think you need a constant fetch

Imagine a user trying to download an hour long 4K video on a slow Internet connection
By default, the fetch will be disrupted by the limitation of the Service Worker's 5-minute lifetime

**Comment by guest271314** [comment]
I don't think it will take 5 minutes to download an hour long 4K video. 

Either of the extension options in https://github.com/guest271314/persistent-serviceworker will keep the MV3 `ServiceWorker` active indefinitely.

**Comment by avi12** [comment]
> Either of the extension options in https://github.com/guest271314/persistent-serviceworker will keep the MV3 ServiceWorker active indefinitely.

Right, but when I tried running  `chromium_extension_web_accessible_resources_iframe_message_event` as-is, it failed to run, so I'm not sure how the code is supposed to be structured to make it possible

> I don't think it will take 5 minutes to download an hour long 4K video.

A 20Mb/s download rate will take a long time, possibly a few hours

**Comment by guest271314** [comment]
> Right, but when I tried running `chromium_extension_web_accessible_resources_iframe_message_event` as-is, it failed to run, so I'm not sure how the code is supposed to be structured to make it possible

That is because you evidently ran the code on a URL beginning with `chrome:`. That won't work. Try on a URL that does not begin with `chrome:` protocol.

**Comment by avi12** [comment]
But how would it look programmatically?
My goal is to fetch the video/audio data on the Service Worker side, not on the content script

**Comment by guest271314** [comment]
It looks the same. On a non-chrome: URL click the icon, an `<iframe>` is appended to the `doument` which exchanges messages with the `ServiceWorker` to keep it active indefinitely, video and audio are requested in the `ServiceWorker`.

**Comment by guest271314** [comment]
That example code has some issues, fixed. 

```
{
const url = "https://www.youtube.com/watch?v=u4nkpIyPkSE";
const text = await (await fetch(url)).text();
const match = text.match(/ytInitialPlayerResponse\s*=\s*({.+?})\s*;\s*(?:var\s+meta|<\/script|\n)/)[1];
const json = JSON.parse(match);
const { adaptiveFormats } = json.streamingData;
const videoObj = adaptiveFormats.find(({ mimeType }) => mimeType.startsWith("video"));
const audioObj = adaptiveFormats.find(({ mimeType }) => mimeType.startsWith("audio"));
const urlVideo = videoObj.url;
const urlAudio = audioObj.url;
const [dataVideo, dataAudio] = await Promise.all([fetch(urlVideo), fetch(urlAudio)]);
console.log(dataVideo, dataAudio);
}
```

**Comment by guest271314** [comment]
I just fetched the audio and video. Do you have an example URL for "an hour long 4K video"?

**Comment by avi12** [comment]
Quick YouTube search gave me [this](https://www.youtube.com/watch?v=wnhvanMdx4s)

**Comment by guest271314** [comment]
What comment did you delete?

**Comment by avi12** [comment]
The comment that fixed my code
I edited my comment to include your fix

**Comment by guest271314** [comment]
Just fetched audio and video in `ServiceWorker`; took a few seconds.

```
// ...
const [dataVideo, dataAudio] = await Promise.all([fetch(urlVideo), fetch(urlAudio)]);
    console.log(dataVideo.headers.get('content-length'), dataAudio.headers.get('content-length'));
}
51990310498 586277884
```

**Comment by guest271314** [comment]
I can record my screen while using the https://github.com/guest271314/persistent-serviceworker/tree/main/chromium_extension_web_accessible_resources_iframe_message_event version, if that will be helpful - though, again, I don't think your issue here is keeping the `ServiceWorker` active over 5 minutes. You are not recording the video in real-time.

**Comment by avi12** [comment]
> Just fetched audio and video in `ServiceWorker`; took a few seconds.

Awesome, I assume your download speed is at least 100Mb/s
But consider users with a poor internet connection who wish to download large videos - their download will need to take at least an hour

**Comment by guest271314** [comment]
Have you tested the persistent `ServiceWorker` code on an _non_-`chrome:` URL?

**Comment by avi12** [comment]
No
I assume you mean to run on a content script, right?

**Comment by guest271314** [comment]
No content script.

Just create an icon to click on and use `executeScript()`. That is what I do to record hours of audio in real-time, from, e.g., the same site.

In pertinent part https://github.com/guest271314/captureSystemAudio/blob/master/native_messaging/capture_system_audio/background.js:

```ts
async nativeMessageStream() {
  return new Promise((resolve) => {
    onmessage = (e) => {
      if (e.origin === this.src.origin) {
        // console.log(e.data);
        if (!this.source) {
          this.source = e.source;
        }
        if (e.data === 1) {
          this.source.postMessage(
            { type: 'start', message: this.stdin },
            '*'
          );
        }
        if (e.data === 0) {
          document.querySelectorAll(`[src="${this.src.href}"]`)
            .forEach((iframe) => {
              document.body.removeChild(iframe);
            });
          onmessage = null;
        }
        if (e.data instanceof ReadableStream) {
          this.stdout = e.data;
          resolve(this.captureSystemAudio());
        }
      }
    };
    this.transferableWindow = document.createElement('iframe');
    this.transferableWindow.style.display = 'none';
    this.transferableWindow.name = location.href;
    this.transferableWindow.src = this.src.href;
    document.body.appendChild(this.transferableWindow);
  }).catch((err) => {
    throw err;
  });
}
```

```ts
chrome.action.onClicked.addListener(async (tab) => {
  if (!(await chrome.storage.local.get('tabId')).tabId) {
    chrome.storage.local.clear();
    await chrome.storage.local.set({ tabId: tab.id });
  }
  chrome.scripting.executeScript({
    target: {
      tabId: (await chrome.storage.local.get('tabId')).tabId,
    },
    world: 'MAIN',
    args: [`${chrome.runtime.getURL('')}transferableStream.html`],
    func: _audioStream,
  });
  if ((await chrome.action.getTitle({})) === 'Capture system audio') {
    await chrome.action.setIcon({ path: './media-playback-stop-1-32x32.png' });
    await chrome.action.setTitle({ title: 'Capturing system audio' });
  } else {
    await chrome.action.setIcon({ path: './media-record-1-32x32.png' });
    await chrome.action.setTitle({ title: 'Capture system audio' });
  }
});
```

Then in the `<iframe>` appended to the `document` - which is a `chrome-extension://<id>` URL that exposes all of the extension functionality in the `ServiceWorker` execpt `onfetch` https://github.com/guest271314/captureSystemAudio/blob/master/native_messaging/capture_system_audio/transferableStream.js

```ts
onload = () => {
  const { readable, writable } = new TransformStream({
    transform(value, controller) {
      controller.enqueue(value);
    },
    flush() {
      console.log('Flush.');
    },
  });
  const writer = writable.getWriter();
  const id = 'capture_system_audio';
  const port = chrome.runtime.connectNative(id);
  port.name = id;
  async function handleMessage(value, port) {
    if (!Array.isArray(value)) {
      value = JSON.parse(value);
    }
    try {
      await writer.ready;
      await writer.write(new Uint8Array(value));
    } catch (e) {
      console.error(e.message);
    }
    return true;
  }
  port.onDisconnect.addListener(async (e) => {
    console.log(e.message);
    await chrome.storage.local.clear();
  });
  port.onMessage.addListener(handleMessage);
  onmessage = async (e) => {
    const { type, message } = e.data;
    if (type === 'start') {
      port.postMessage(message);
      parent.postMessage(readable, name, [readable]);
    }
    if (type === 'stop') {
      try {
        port.disconnect(id);
        console.log(writer.desiredSize, message);
        while (writer.desiredSize <= 1) {    
          await scheduler.postTask(() => {});
          if (writer.desiredSize === 1) {
            console.log('writable', writer.desiredSize);   
            break;
          }
        }
        await writer.close();
        await writer.closed;
        await writer.close();
        await writer.closed;
        console.log(writer.desiredSize);
        parent.postMessage(0, name);
        onmessage = null;
        await chrome.storage.local.clear();
      } catch (err) {
        console.error(err.message);
      }
    }
  };
  parent.postMessage(1, name);
};
```

**Comment by guest271314** [comment]
Basically, for your use case you really don't need the `ServiceWorker` to stay active indefinitely. Just append an `<iframe>` to the clicked `document` - that is not a `chrome:` URL - all of the chrome.* API's will be exposed in the `iframe`, without any time restriction applied whatsoever.

However, if you really think you need an MV3 `ServiceWorker` to remain active indefinitely, you can do that too. I suspect the users won't be clicking the icon on the `chrome://extensions` URL - though if you want that functionality as well, you can query the open tabs for the URL that does not begin with `chrome:`, append the `iframe` to one of those Tab's, and when the download and encoding are complete use Notifications API and File System Access API to notify user file is ready to download and write the file directly to the users' file system, e.g., 

```ts
audioStream = new AudioStream('parec -d @DEFAULT_MONITOR@', 'audio/webm;codecs=opus');
// audioStream.mediaStream: live MediaStream
audioStream
  .start()
  .then(async (ab) => {
    // ab: ArrayBuffer representation of WebM file from MediaRecorder
    console.log(audioStream);
    const {
      src: { origin },
      mimeType,
    } = audioStream;
    const blob = new Blob([ab], { type: mimeType });
    console.log(
      URL.createObjectURL(blob),
      `recording${new Date().getTime()}${
        mimeType.includes('opus') ? '.webm' : '.mp3'
      }`
    );

    globalThis.audioStream = AudioStream = null;
    let permission = await navigator.permissions.query({
      name: 'notifications',
    });
    if (permission.state !== 'granted') {
      permission = await Notification.requestPermission();
    }
    if (permission.state === 'granted' || permission === 'granted') {
      const saveFileNotification = new Notification('Save file?', {
        body: `Click "Activate" to download captured system audio recording.`,
        icon: `${origin}/download_music_icon.png`,
      });
      saveFileNotification.onclick = async (e) => {
        try {
          const handle = await showSaveFilePicker({
            startIn: 'music',
            suggestedName: `recording${new Date().getTime()}${
              mimeType.includes('opus') ? '.webm' : '.mp3'
            }`,
          });
          console.log(handle);
          const writable = await handle.createWritable();
          const writer = writable.getWriter();
          await writer.write(blob);
          await writer.close();
        } catch (err) {
          console.error(err);
        }
      };
    }
  })
  .catch(console.error);
```

**Comment by guest271314** [comment]
Keep in mind the HTML loaded in the `<iframe>` is in the extension, accessible due to 

```json
"web_accessible_resources": [{
  "resources": [ "*.html", "*.js", "*.png" ],
  "matches": [ "<all_urls>" ],
  "extensions": [ ]
}]
```

**Comment by avi12** [comment]
OH now I got it
So technically, I could have a listener on the Service Worker that will be triggered by a button that is injected into YouTube

Something like
`content-script.js`
```ts
new MutationObserver((_, observer) => {
  const elButtonContainer = document.querySelector(someSelector);
  elButtonContainer.insertAdjacentHTML("afterbegin", `<button id="download">Download</button>`);
}).observe(document. { childList: true, subtree: true });

document.addEventListener("click", e => {
  const elButton = e.target;
  if (elButton?.id !== "download") {
    return;
  }

  chrome.runtime.sendMessage({ urlYouTube: location.href })
});
```
`background.js`
```ts
chrome.runtime.onMessage.addListener(async (data, { tab }) => {
    if (!currentPersistentClient) {
    currentPersistentClient = tab.url;
    await chrome.scripting.executeScript({
      target: {
        tabId: tab.id,
      },
      world: 'MAIN',
      args: [chrome.runtime.getURL('')],
      func: persistServiceWorker,
    });
    (
      await self.clients.matchAll({ includeUncontrolled: true })
    )[0].postMessage(data, [port2]);
  } else {
    await chrome.scripting.executeScript({
      target: {
        tabId: tab.id,
      },
      world: 'MAIN',
      args: [currentPersistentClient],
      func: persistentServiceWorkerActive,
    });
  }
});
```
`index.js`
```ts
async function getData(url) {
  const text = await (await fetch(url)).text();
  const match = text.match(/ytInitialPlayerResponse\s*=\s*({.+?})\s*;\s*(?:var\s+meta|<\/script|\n)/)[1];
  const json = JSON.parse(match);
  const { adaptiveFormats } = json.streamingData;
  const videoObj = adaptiveFormats.find(({ mimeType }) => mimeType.startsWith("video"));
  const audioObj = adaptiveFormats.find(({ mimeType }) => mimeType.startsWith("audio"));
  const urlVideo = videoObj.url;
  const urlAudio = audioObj.url;

  const [dataVideo, dataAudio] = await Promise.all([fetch(urlVideo), fetch(urlAudio)]);
  return [dataVideo, dataAudio];
}

navigator.serviceWorker.addEventListener("message", async e => {
  parent.postMessage(e.data, name, e.ports);
  try {
    while ((await navigator.serviceWorker.ready).active) {
      (await navigator.serviceWorker.ready).active.postMessage(await getData(e.data.urlYouTube));
      await new Promise((resolve) => setTimeout(resolve, 1000 * 15));
    }
    closeMessageChannel();
  } catch (err) {
    console.error(err);
    closeMessageChannel();
  }
});
```

**Comment by guest271314** [comment]
I think you will need the `iframe` with `src` set the the `chrome-extension:` URL to establish a `WindowClient` for the `ServiceWorker`.
